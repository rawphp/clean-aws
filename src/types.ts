import { EventEmitter } from 'events';

export interface IResourceCleaner extends EventEmitter {
  list: () => Promise<object>;
  remove: (list: IRegionResources) => Promise<number>;
  toString: () => string;
}

export interface ICleanOptions {
  profile: string;
  region: string;
  resourceFile: string;
  dryRun: boolean;
  [key: string]: any;
}

// document types

export type ValidRegion = 'ap-southeast-2' | 'eu-west-1';

export interface IMasterResourceList {
  'us-east-1'?: IRegionResources;
  'us-east-2'?: IRegionResources;
  'us-west-1'?: IRegionResources;
  'us-west-2'?: IRegionResources;

  'ap-south-1'?: IRegionResources;

  'ap-northeast-1'?: IRegionResources;
  'ap-northeast-2'?: IRegionResources;
  'ap-northeast-3'?: IRegionResources;
  'ap-southeast-1'?: IRegionResources;
  'ap-southeast-2'?: IRegionResources;
  'ca-central-1'?: IRegionResources;
  'cn-north-1'?: IRegionResources;
  'cn-northwest-1'?: IRegionResources;

  'eu-central-1'?: IRegionResources;
  'eu-west-1'?: IRegionResources;
  'eu-west-2'?: IRegionResources;
  'eu-west-3'?: IRegionResources;

  'sa-east-1'?: IRegionResources;
  'us-gov-east-1'?: IRegionResources;
  'us-gov-west-1'?: IRegionResources;
}

export type IRegionResources =
  & ISQSList
  & ISNSList
  & IBucketsList
  & IIAMList
  & IEC2List
  & IDynamoDbTableList
  & IDataPipelineList
  & ICloudWatchList
  & ICloudFormationList;

export interface ISQSList {
  region: string;
  profile: string;
  queues: string[];
}

export interface ISNSList {
  region: string;
  profile: string;
  topics: string[];
  subscriptions: string[];
}

export interface IBucketsList {
  region: string;
  profile: string;
  buckets: string[];
}

export interface IIAMList {
  region: string;
  profile: string;
  policies: string[];
  roles: string[];
  accessKeys: string[];
  users: string[];
}

export interface IEC2List {
  region: string;
  profile: string;
  ec2s: string[];
}

export interface IDynamoDbTableList {
  region: string;
  profile: string;
  tables: string[];
}

export interface IDataPipelineList {
  region: string;
  profile: string;
  dataPipelines: string[];
}

export interface ICloudWatchList {
  region: string;
  profile: string;
  logGroups: string[];
}

export interface ICloudFormationList {
  region: string;
  profile: string;
  cloudFormationStacks: string[];
}
