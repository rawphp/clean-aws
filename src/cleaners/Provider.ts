import { EventEmitter } from 'events';
import { IProviderOptions } from '../types';

export abstract class Provider extends EventEmitter {
  public region: string;
  public profile: string;
  public dryRun: boolean;
  protected name: string;

  public constructor(name: string, options: IProviderOptions) {
    super();

    this.name = name;
    this.region = options.region;
    this.profile = options.profile;
    this.dryRun = options.dryRun || false;
  }

  public toString() {
    return `${this.name}-${this.profile}-${this.region}`;
  }
}
