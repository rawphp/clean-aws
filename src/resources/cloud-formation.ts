import * as AWS from 'aws-sdk';
import { ICleanOptions, IResourceCleaner } from '../types';

export interface ICloudFormationList {
  region: string;
  cloudFormationStacks: string[];
}

export class CloudFormation implements IResourceCleaner {
  protected cloudFormation: AWS.CloudFormation;
  protected region: string;

  public constructor(options: ICleanOptions) {
    this.region = options.region;

    this.cloudFormation = new AWS.CloudFormation({
      region: options && options.region ? options.region : '',
    });
  }

  public async list(): Promise<object> {
    console.log('[CloudFormation] Listing Stacks', this.region);

    try {
      let next: boolean | string = true;

      const file: ICloudFormationList = {
        region: this.region,
        cloudFormationStacks: [],
      };

      while (next) {
        const stacks = await this.cloudFormation.describeStacks().promise();

        if (stacks && stacks.Stacks) {
          stacks.Stacks.forEach((stack: AWS.CloudFormation.Stack) => {
            file.cloudFormationStacks.push(stack.StackName);
          });
        }

        next = false;

        if (stacks.NextToken) {
          next = stacks.NextToken;
        }
      }

      return file;
    } catch (error) {
      console.error(error);

      throw error;
    }
  }

  public async remove(stacks: string[]): Promise<number> {
    try {
      const processes = stacks.map((stack: string) => {
        return this.cloudFormation
          .deleteStack({
            StackName: stack,
          })
          .promise();
      });

      await Promise.all(processes);

      return 0;
    } catch (error) {
      console.error(error);

      return 1;
    }
  }
}
