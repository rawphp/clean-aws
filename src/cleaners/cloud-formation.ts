import * as AWS from 'aws-sdk';
import * as BPromise from 'bluebird';
import * as colors from 'colors/safe';
import { isBoolean } from 'util';
import { IProviderOptions, IResourceCleaner } from '../types';
import { Provider } from './Provider';

export class CloudFormation extends Provider implements IResourceCleaner {
  protected cloudFormation: AWS.CloudFormation;
  protected stacks: string[];

  public constructor(options: IProviderOptions) {
    super('CloudFormation', options);

    this.stacks = options && options.stacks || [];
    this.cloudFormation = options.cloudFormation || new AWS.CloudFormation({
      region: this.region,
    });
  }

  public getData(): any {
    return this.stacks;
  }

  public async list(): Promise<boolean> {
    console.log('[CloudFormation] Listing Stacks', this.region);

    this.emit('listStarted', { resource: this, region: this.region });

    try {
      let next: boolean | string | undefined = true;

      while (next) {
        if (isBoolean(next)) {
          next = undefined;
        }

        const stacks = await this.cloudFormation.describeStacks({
          NextToken: next as string,
        }).promise();

        if (stacks && stacks.Stacks) {
          stacks.Stacks.forEach((stack: AWS.CloudFormation.Stack) => {
            this.stacks.push(stack.StackName);
          });
        }

        next = false;

        if (stacks.NextToken) {
          console.log('has NextToken', stacks.NextToken);

          next = stacks.NextToken;
        }
      }

      this.emit('listCompleted', { resource: this, region: this.region, data: { stacks: this.stacks } });

      return true;
    } catch (error) {
      console.error(error);

      throw error;
    }
  }

  public async remove(): Promise<boolean> {
    console.log('[CloudFormation] Deleting Stacks');

    try {
      this.emit('deleteStarted', { resource: this, region: this.region, data: { stacks: this.stacks } });

      if (this.dryRun) {
        console.log('[CloudFormation] Dry Run');

        return true;
      }

      await BPromise.map(this.stacks, async (stack: string) => {
        try {
          await this.cloudFormation.deleteStack({
            StackName: stack,
          }).promise();

          // poll completion
          let complete: boolean = false;

          do {
            try {
              await BPromise.delay(1000 * 2);

              const exists = await this.cloudFormation.describeStacks({
                StackName: stack,
              }).promise();

              complete = exists.Stacks && exists.Stacks.length > 0 ? false : true;
            } catch (error) {
              console.error(colors.yellow(error.message));

              complete = true;
            }
          } while (!complete);

        } catch (error) {
          console.error(colors.red(error.message));
        }
      }, { concurrency: 4 });

      this.emit('deleteCompleted', { resource: this, region: this.region, data: { stacks: this.stacks } });

      return true;
    } catch (error) {
      console.error(colors.red(error.message));

      return false;
    }
  }
}
