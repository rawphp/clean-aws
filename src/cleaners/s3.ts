import * as AWS from 'aws-sdk';
import * as BPromise from 'bluebird';
import * as colors from 'colors/safe';
import { IResourceCleaner, IS3Options } from '../types';
import { Provider } from './Provider';

export class S3 extends Provider implements IResourceCleaner {
  protected s3: AWS.S3;
  protected buckets: string[];

  public constructor(options: IS3Options) {
    super('S3', options);

    this.buckets = options.buckets || [];
    this.s3 = options.s3 || new AWS.S3({
      region: options && options.region ? options.region : undefined,
    });
  }

  public getData(): any {
    return this.buckets;
  }

  public async list(): Promise<boolean> {
    console.log('[S3] Listing Buckets');

    this.emit('listStarted', { resource: this, region: this.region });

    try {
      const response = await this.s3.listBuckets().promise();

      if (!response || !response.Buckets) {
        return false;
      }

      this.buckets = response && response.Buckets ? response.Buckets.map((bucket: AWS.S3.Bucket) => bucket.Name as string) : [];

      this.emit('listCompleted', { resource: this, region: this.region, data: { buckets: this.buckets } });

      return true;
    } catch (error) {
      console.error(colors.red(error.message));

      return false;
    }
  }

  public async remove(): Promise<boolean> {
    console.log('[S3] Deleting Buckets');

    try {
      this.emit('deleteStarted', { resource: this, region: this.region, data: { buckets: this.buckets } });

      if (this.dryRun) {
        console.log('[S3] Dry Run');

        return true;
      }

      if (!this.buckets) {
        throw new Error('Bucket list has not been set');
      }

      await BPromise.map(this.buckets, async (bucket: string) => {
        try {
          let token: string | undefined;
          let versionMarker: string | undefined;
          let keyMarker: string | undefined;

          this.emit('deletionStarted', { region: this.region, name: bucket });

          do {
            const listResponse: AWS.S3.ListObjectsV2Output = await this.s3.listObjectsV2({
              Bucket: bucket,
              ContinuationToken: token,
            }).promise();

            // delete objects
            await BPromise.map(listResponse.Contents || [], async (obj) => {
              try {
                await this.s3.deleteObject({
                  Bucket: bucket,
                  Key: obj.Key,
                } as AWS.S3.DeleteObjectRequest).promise();
                process.stdout.write('.');
              } catch (error) {
                console.error(error.message);
              }
            });

            token = listResponse.NextContinuationToken;
          } while (token);

          do {
            const listVersionsResponse = await this.s3.listObjectVersions({
              Bucket: bucket,
              KeyMarker: keyMarker,
              VersionIdMarker: versionMarker,
            }).promise();

            // delete versions
            await BPromise.map(listVersionsResponse.Versions || [], async (obj) => {
              try {
                await this.s3.deleteObject({
                  Bucket: bucket,
                  Key: obj.Key,
                  VersionId: obj.VersionId,
                } as AWS.S3.DeleteObjectRequest).promise();
                process.stdout.write('.');
              } catch (error) {
                console.error(error.message);
              }
            }, { concurrency: 1000 });

            // delete delete markers
            await BPromise.map(listVersionsResponse.DeleteMarkers || [], async (obj) => {
              try {
                await this.s3.deleteObject({
                  Bucket: bucket,
                  Key: obj.Key,
                  VersionId: obj.VersionId,
                } as AWS.S3.DeleteObjectRequest).promise();
                process.stdout.write('.');
              } catch (error) {
                console.error(error.message);
              }
            }, { concurrency: 1000 });

            versionMarker = listVersionsResponse.NextVersionIdMarker;
            keyMarker = listVersionsResponse.KeyMarker;
          } while (versionMarker);

          try {
            // delete bucket
            await this.s3.deleteBucket({
              Bucket: bucket,
            }).promise();
            process.stdout.write('.');
          } catch (error) {
            console.error(colors.red('Delete bucket error'), bucket, error.message);
          }

          this.emit('deletionCompleted', { region: this.region, name: bucket });
        } catch (error) {
          console.error(colors.red(error.message));
        }
      }, { concurrency: 5 });

      this.emit('deleteCompleted', { resource: this, region: this.region, data: this.buckets });

      return true;
    } catch (error) {
      if (error.message === 'The specified bucket does not exist') {
        return true;
      }

      console.error(colors.red(error.message));

      return false;
    }
  }
}
