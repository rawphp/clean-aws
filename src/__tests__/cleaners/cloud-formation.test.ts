import * as AWS from 'aws-sdk';
import * as nock from 'nock';
import { CloudFormation } from '../../cleaners/cloud-formation';
import { IProviderOptions, IResourceCleaner } from '../../types';
import * as describeStacksResponse from '../fixtures/describe-stacks-response.json';

describe('CloudFormation', () => {
  let resource: IResourceCleaner;
  let cf: AWS.CloudFormation;

  const options: IProviderOptions = {
    profile: 'default',
    region: 'ap-southeast-2',
    dryRun: false,
    resourceFile: 'resources.json',
  };

  beforeEach(() => {
    nock.disableNetConnect();
    // nock('https://cloudformation.ap-southeast-2.amazonaws.com').post('/').reply(200, {});

    cf = new AWS.CloudFormation({ region: options.region });

    options.cloudFormation = cf;

    cf.describeStacks = jest.fn().mockReturnValueOnce({
      promise: () => describeStacksResponse,
    });

    resource = new CloudFormation(options);
  });

  afterEach(() => {
    nock.enableNetConnect();
    jest.restoreAllMocks();
  });

  describe('toString', () => {
    it('toString returns correct value', () => {
      expect(resource.toString()).toEqual('CloudFormation-default-ap-southeast-2');
    });
  });

  describe('list', () => {
    it('lists stacks successfully', async () => {
      await resource.list();

      expect(resource.getData()).toMatchSnapshot();
    });

    it('emits listStarted event', async () => {
      let started = false;

      resource.on('listStarted', (data) => {
        started = true;
      });

      await resource.list();

      expect(started).toEqual(true);
    });

    it('emits listCompleted event', async () => {
      let completed = false;

      resource.on('listCompleted', (data) => {
        completed = true;
      });

      await resource.list();

      expect(completed).toEqual(true);
    });
  });

  describe('remove', () => {
    it('skips non-existent stacks', async () => {
      const stacks = ['tdk-test-stack'];

      resource = new CloudFormation({
        ...options,
        stacks,
      });

      expect(await resource.remove()).toEqual(true);
    });

    it('successfully removes a stack', async () => {
      const stacks = ['ups-v3-tdk'];

      resource = new CloudFormation({
        ...options,
        stacks,
      });

      cf.deleteStack = jest.fn().mockReturnValueOnce({
        promise: () => ({
          ResponseMetadata: {
            RequestId: '49f2b134-8bd4-11e9-bdf1-519eaf8997bd',
          },
        }),
      });

      cf.describeStacks = jest
        .fn()
        .mockReturnValueOnce({
          promise: () => ({
            ResponseMetadata: { RequestId: '4b559c28-8bd4-11e9-85bd-c122c3e7f8c1' },
            Stacks: [
              {
                // tslint:disable-next-line:max-line-length
                StackId: 'arn:aws:cloudformation:ap-southeast-2:159931620759:stack/ups-v3-tdk/a165c3f0-f2df-11e8-aa1d-0ab1e239639c',
                StackName: 'ups-v3-tdk',
                Description: 'The AWS CloudFormation template for this Serverless application',
                Parameters: [],
                CreationTime: '2018-11-28T07:31:33.431Z',
                DeletionTime: '2019-06-10T23:05:50.259Z',
                LastUpdatedTime: '2019-03-05T03:30:01.815Z',
                RollbackConfiguration: {},
                StackStatus: 'DELETE_IN_PROGRESS',
                DisableRollback: false,
                NotificationARNs: [],
                Capabilities: ['CAPABILITY_IAM', 'CAPABILITY_NAMED_IAM'],
                Outputs: [
                  {
                    OutputKey: 'MarcoLambdaFunctionQualifiedArn',
                    OutputValue: 'arn:aws:lambda:ap-southeast-2:159931620759:function:ups-v3-tdk-marco:15',
                    Description: 'Current Lambda function version',
                  },
                  {
                    OutputKey: 'TrackingLambdaFunctionQualifiedArn',
                    OutputValue: 'arn:aws:lambda:ap-southeast-2:159931620759:function:ups-v3-tdk-tracking:15',
                    Description: 'Current Lambda function version',
                  },
                  {
                    OutputKey: 'BookingLambdaFunctionQualifiedArn',
                    OutputValue: 'arn:aws:lambda:ap-southeast-2:159931620759:function:ups-v3-tdk-booking:15',
                    Description: 'Current Lambda function version',
                  },
                  {
                    OutputKey: 'RegisterLambdaFunctionQualifiedArn',
                    OutputValue: 'arn:aws:lambda:ap-southeast-2:159931620759:function:ups-v3-tdk-register:15',
                    Description: 'Current Lambda function version',
                  },
                  {
                    OutputKey: 'PaperlessLambdaFunctionQualifiedArn',
                    OutputValue: 'arn:aws:lambda:ap-southeast-2:159931620759:function:ups-v3-tdk-paperless:15',
                    Description: 'Current Lambda function version',
                  },
                  {
                    OutputKey: 'QuoteLambdaFunctionQualifiedArn',
                    OutputValue: 'arn:aws:lambda:ap-southeast-2:159931620759:function:ups-v3-tdk-quote:15',
                    Description: 'Current Lambda function version',
                  },
                  // tslint:disable-next-line:max-line-length
                  { OutputKey: 'ServerlessDeploymentBucketName', OutputValue: 'ups-v3-tdk-serverlessdeploymentbucket-f7h6ngs32no' },
                  {
                    OutputKey: 'MetaLambdaFunctionQualifiedArn',
                    OutputValue: 'arn:aws:lambda:ap-southeast-2:159931620759:function:ups-v3-tdk-meta:15',
                    Description: 'Current Lambda function version',
                  },
                  {
                    OutputKey: 'IamRoleArnLambda',
                    OutputValue: 'arn:aws:iam::159931620759:role/ups-v3-tdk-IamRoleLambda-17FY68BTVCPHY',
                    Description: 'ARN of the lambda IAM role',
                  },
                  {
                    OutputKey: 'LicenseLambdaFunctionQualifiedArn',
                    OutputValue: 'arn:aws:lambda:ap-southeast-2:159931620759:function:ups-v3-tdk-license:15',
                    Description: 'Current Lambda function version',
                  },
                  {
                    OutputKey: 'EulaLambdaFunctionQualifiedArn',
                    OutputValue: 'arn:aws:lambda:ap-southeast-2:159931620759:function:ups-v3-tdk-eula:15',
                    Description: 'Current Lambda function version',
                  },
                  {
                    OutputKey: 'CompletionLambdaFunctionQualifiedArn',
                    OutputValue: 'arn:aws:lambda:ap-southeast-2:159931620759:function:ups-v3-tdk-completion:15',
                    Description: 'Current Lambda function version',
                  },
                  {
                    OutputKey: 'CollectionPointLambdaFunctionQualifiedArn',
                    OutputValue: 'arn:aws:lambda:ap-southeast-2:159931620759:function:ups-v3-tdk-collectionPoint:15',
                    Description: 'Current Lambda function version',
                  },
                  {
                    OutputKey: 'ServiceEndpoint',
                    OutputValue: 'https://yxsgif1eeh.execute-api.ap-southeast-2.amazonaws.com/tdk',
                    Description: 'URL of the service endpoint',
                  },
                ],
                Tags: [{ Key: 'STAGE', Value: 'tdk' }],
                EnableTerminationProtection: false,
                DriftInformation: { StackDriftStatus: 'NOT_CHECKED' },
              },
            ],
          }),
        })
        .mockReturnValueOnce({
          promise: () => {
            throw new Error('problem occurred');
          },
        });

      expect(await resource.remove()).toEqual(true);
    });
  });
});
