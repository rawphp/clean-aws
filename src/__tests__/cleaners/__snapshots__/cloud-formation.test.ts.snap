// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CloudFormation list lists stacks successfully 1`] = `
Array [
  "test-document-service-v2-abc",
  "test-scheduled-reports-abc",
  "carrier-integration-stub-abc",
  "ups-v3-abc",
  "collection-point-abc",
  "test-nginx-proxy-abc",
  "test-completion-service-abc",
  "test-experience-service-abc",
  "test-carrier-service-abc",
  "tracking-role-service-abc",
  "test-shipment-service-abc",
  "test-account-service-abc",
  "test-order-service-abc",
  "awseb-e-umysvmxihi-stack",
  "schemas-service-abc",
  "providence-auth-abc",
  "credentials-abc-support",
  "dpd-uk-abc",
  "credentials-abc-carriers",
  "test-document-service-abc",
  "test-tracking-service-abc",
  "test-product-service-abc",
  "test-rate-service-abc",
  "rate-templates-abc",
  "test-session-service-abc",
  "test-documentation-storage-abc",
  "test-container-service-abc",
  "packing-service-abc",
  "test-stream-service-abc",
  "magento-shipping-service-abc",
  "test-notification-service-abc",
  "collection-point-service-abc",
  "test-location-service-abc",
  "test-billing-service-abc",
  "test-sequencer-abc",
  "credentials-abc-phoenix",
]
`;
